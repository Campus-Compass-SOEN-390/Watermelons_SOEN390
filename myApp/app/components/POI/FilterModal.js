// src/components/FilterModal.js
import React, { useContext } from "react";
import {
  Modal,
  View,
  Text,
  TouchableOpacity,
  Switch,
  SafeAreaView,
} from "react-native";
import Slider from "@react-native-community/slider";
import PropTypes from "prop-types";
import { styles as defaultStyles } from "../../styles/poiStyles";
import { ThemeContext } from "../../context/ThemeContext";

// Create theme-aware styles for the FilterModal
const getThemedStyles = (isDarkMode, theme, styles) => {
  // If we have theme and isDarkMode props, create dynamic styles
  if (theme && isDarkMode !== undefined) {
    const backgroundColor = isDarkMode
      ? theme.cardBackground || "#1e1e1e"
      : "#fff";
    const textColor = isDarkMode ? theme.text || "#fff" : "#333";
    const buttonColor = theme.buttonBackground || "#922338";

    return {
      ...styles,
      filterModalContainer: {
        ...styles.filterModalContainer,
        backgroundColor: "rgba(0,0,0,0.6)", // Slightly darker overlay in dark mode
      },
      filterModalContent: {
        ...styles.filterModalContent,
        backgroundColor: backgroundColor,
        shadowColor: "#000",
        shadowOpacity: isDarkMode ? 0.4 : 0.2,
      },
      filterModalTitle: {
        ...styles.filterModalTitle,
        color: textColor,
      },
      filterSectionTitle: {
        ...styles.filterSectionTitle,
        color: textColor,
      },
      filterOptionText: {
        ...styles.filterOptionText,
        color: textColor,
      },
      distanceRangeLabel: {
        ...styles.distanceRangeLabel,
        color: textColor,
      },
      closeModalIcon: {
        ...styles.closeModalIcon,
        color: textColor,
      },
      applyButton: {
        ...styles.applyButton,
        backgroundColor: buttonColor,
      },
    };
  }

  // If no theme info provided, return default styles
  return styles;
};

const FilterModal = ({
  isVisible,
  onClose,
  distance,
  setDistance,
  showCafes,
  setShowCafes,
  showRestaurants,
  setShowRestaurants,
  showActivities,
  setShowActivities,
  useDistance,
  setUseDistance,
  isDarkMode,
  theme,
}) => {
  // If theme/isDarkMode weren't passed as props, get from context
  const themeContext = useContext(ThemeContext);
  const effectiveIsDarkMode =
    isDarkMode !== undefined ? isDarkMode : themeContext?.isDarkMode;
  const effectiveTheme = theme || themeContext?.theme;

  // Get appropriate styles based on theme
  const themedStyles = getThemedStyles(
    effectiveIsDarkMode,
    effectiveTheme,
    defaultStyles
  );

  // Theme-aware switch colors
  const trackColor = {
    false: effectiveIsDarkMode ? "#555" : "#e0e0e0",
    true: effectiveIsDarkMode ? "#7a3741" : "#d6909a",
  };
  const thumbColor = (value) => {
    if (effectiveIsDarkMode) {
      return value ? "#922338" : "#999";
    }
    return value ? "#922338" : "#f4f3f4";
  };

  return (
    <Modal
      visible={isVisible}
      transparent
      animationType="slide"
      onRequestClose={onClose}
    >
      <View style={themedStyles.filterModalContainer}>
        <SafeAreaView>
          <View style={themedStyles.filterModalContent}>
            {/* Header with title and close button */}
            <View style={themedStyles.filterModalHeader}>
              <Text style={themedStyles.filterModalTitle}>
                Filter Points of Interest
              </Text>
              <TouchableOpacity
                style={themedStyles.closeModalButton}
                onPress={onClose}
              >
                <Text style={themedStyles.closeModalIcon}>Ã—</Text>
              </TouchableOpacity>
            </View>

            {/* Distance Section */}
            <View style={themedStyles.filterSection}>
              <Text style={themedStyles.filterSectionTitle}>
                Distance Filter
              </Text>

              <View style={themedStyles.filterOption}>
                <Text style={themedStyles.filterOptionText}>
                  Enable distance filtering
                </Text>
                <Switch
                  testID="distanceSwitch"
                  value={useDistance}
                  onValueChange={setUseDistance}
                  trackColor={trackColor}
                  thumbColor={thumbColor(useDistance)}
                  ios_backgroundColor={effectiveIsDarkMode ? "#555" : "#e0e0e0"}
                />
              </View>

              {useDistance && (
                <View style={themedStyles.distanceSliderContainer}>
                  <Text style={themedStyles.distanceRangeLabel}>
                    Maximum distance: {distance} kilometers
                  </Text>
                  <Slider
                    style={themedStyles.customSlider}
                    minimumValue={0.5}
                    maximumValue={10}
                    step={0.5}
                    value={distance}
                    onValueChange={setDistance}
                    minimumTrackTintColor="#922338"
                    maximumTrackTintColor={
                      effectiveIsDarkMode ? "#555" : "#e0e0e0"
                    }
                    thumbTintColor="#922338"
                  />
                </View>
              )}
            </View>

            {/* Categories Section */}
            <View style={themedStyles.filterSection}>
              <Text style={themedStyles.filterSectionTitle}>Categories</Text>

              <View style={themedStyles.filterOption}>
                <Text style={themedStyles.filterOptionText}>Cafes</Text>
                <Switch
                  testID="cafeSwitch"
                  value={showCafes}
                  onValueChange={setShowCafes}
                  trackColor={trackColor}
                  thumbColor={thumbColor(showCafes)}
                  ios_backgroundColor={effectiveIsDarkMode ? "#555" : "#e0e0e0"}
                />
              </View>

              <View style={themedStyles.filterOption}>
                <Text style={themedStyles.filterOptionText}>Restaurants</Text>
                <Switch
                  testID="restaurantSwitch"
                  value={showRestaurants}
                  onValueChange={setShowRestaurants}
                  trackColor={trackColor}
                  thumbColor={thumbColor(showRestaurants)}
                  ios_backgroundColor={effectiveIsDarkMode ? "#555" : "#e0e0e0"}
                />
              </View>

              <View style={themedStyles.filterOption}>
                <Text style={themedStyles.filterOptionText}>Activities</Text>
                <Switch
                  testID="activitySwitch"
                  value={showActivities}
                  onValueChange={setShowActivities}
                  trackColor={trackColor}
                  thumbColor={thumbColor(showActivities)}
                  ios_backgroundColor={effectiveIsDarkMode ? "#555" : "#e0e0e0"}
                />
              </View>
            </View>

            {/* Apply Button */}
            <TouchableOpacity
              style={themedStyles.applyButton}
              onPress={onClose}
            >
              <Text style={themedStyles.applyButtonText}>Apply Filters</Text>
            </TouchableOpacity>
          </View>
        </SafeAreaView>
      </View>
    </Modal>
  );
};

FilterModal.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  distance: PropTypes.number.isRequired,
  setDistance: PropTypes.func.isRequired,
  showCafes: PropTypes.bool.isRequired,
  setShowCafes: PropTypes.func.isRequired,
  showRestaurants: PropTypes.bool.isRequired,
  setShowRestaurants: PropTypes.func.isRequired,
  showActivities: PropTypes.bool.isRequired,
  setShowActivities: PropTypes.func.isRequired,
  useDistance: PropTypes.bool.isRequired,
  setUseDistance: PropTypes.func.isRequired,
  // Add new theme props
  isDarkMode: PropTypes.bool,
  theme: PropTypes.object,
};

export default FilterModal;
