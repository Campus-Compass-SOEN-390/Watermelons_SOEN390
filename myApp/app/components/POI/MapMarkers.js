import React, { useContext } from "react";
import Mapbox from "@rnmapbox/maps";
import { TouchableOpacity } from "react-native";
import { getRandomBytes } from "expo-crypto";
import { ThemeContext } from "../../context/ThemeContext"; // Update path if needed

// Update MapMarkers to handle coordinates correctly, make markers clickable, and support themes
export default function MapMarkers({
  data,
  MarkerComponent,
  onMarkerPress,
  isDarkMode: propIsDarkMode,
  theme: propTheme,
}) {
  // Get theme from context if not provided as props
  const themeContext = useContext(ThemeContext);
  const isDarkMode =
    propIsDarkMode !== undefined ? propIsDarkMode : themeContext?.isDarkMode;
  const theme = propTheme || themeContext?.theme;

  // Memoize the theme context value to avoid unnecessary re-renders
  const themeContextValue = React.useMemo(
    () => ({ isDarkMode, theme }),
    [isDarkMode, theme]
  );

  // Pass theme info to MarkerComponent if it needs it
  const MarkerWithTheme = (props) => (
    <ThemeContext.Provider value={themeContextValue}>
      <MarkerComponent {...props} />
    </ThemeContext.Provider>
  );

  console.log("MapMarkers data:", data[0]);
  return data.map((point) => {
    const lat = point?.geometry?.location?.lat;
    const lng = point?.geometry?.location?.lng;
    if (!lat || !lng) return null;

    // Generate a unique key/id using place_id or fallback to a random id
    // Secure random ID generator -- sonarqube imporvement
    const generateSecureId = () => {
      const randomBytes = getRandomBytes(8); // 8 bytes = 16 hex chars
      return Array.from(randomBytes, (byte) =>
        byte.toString(16).padStart(2, "0")
      ).join("");
    };

    // Usage:
    const pointId = point.place_id || `marker-${generateSecureId()}`;

    return (
      <Mapbox.MarkerView
        key={pointId}
        id={pointId}
        coordinate={[lng, lat]}
        title={point.name || "Location"}
        onPress={() => onMarkerPress?.(point)}
      >
        <TouchableOpacity onPress={() => onMarkerPress?.(point)}>
          <MarkerWithTheme />
        </TouchableOpacity>
      </Mapbox.MarkerView>
    );
  });
}
