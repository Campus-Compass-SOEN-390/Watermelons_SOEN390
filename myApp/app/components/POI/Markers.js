// src/components/Markers.js
import React from "react";
import { View } from "react-native";
import { MaterialCommunityIcons } from "@expo/vector-icons";

// Custom marker components
const CoffeeMarker = () => (
  <View style={{ backgroundColor: 'white', padding: 5, borderRadius: 5 }}>
    <MaterialCommunityIcons name="coffee" size={24} color="black" />
  </View>
);

const RestaurantMarker = () => (
  <View style={{ backgroundColor: 'white', padding: 5, borderRadius: 5 }}>
    <MaterialCommunityIcons name="silverware-fork-knife" size={24} color="orange" />
  </View>
);

const ActivityMarker = () => (
  <View style={{ backgroundColor: 'white', padding: 5, borderRadius: 5 }}>
    <MaterialCommunityIcons name="run" size={20} color="green" />
  </View>
);

export { CoffeeMarker, RestaurantMarker, ActivityMarker };