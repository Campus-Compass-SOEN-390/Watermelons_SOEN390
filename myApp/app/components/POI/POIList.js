import React, { useState, useContext, useCallback, memo, useMemo, useRef } from "react";
import PropTypes from "prop-types";
import {
    View,
    Text,
    Image,
    FlatList,
    ActivityIndicator,
    TouchableOpacity,
    RefreshControl,
    Platform,
    Animated
} from "react-native";
import { Ionicons } from "@expo/vector-icons";
import {
  styles as defaultStyles,
  createPOIListStyles,
} from "../../styles/POIListStyle";
import Constants from "expo-constants";
import { useRouter } from "expo-router";
import { useLocationContext } from "@/app/context/LocationContext";
import { ThemeContext } from "@/app/context/ThemeContext";
import { useButtonInteraction } from "../../hooks/useButtonInteraction";

const GOOGLE_PLACES_API_KEY = Constants.expoConfig?.extra?.apiKey;

// Constants for consistent item heights (important for smooth scrolling)
const ITEM_HEIGHT = 350; // Base height including image, content, padding
const IMAGE_HEIGHT = 180; // Match the style definition
const SCROLL_THRESHOLD = 300; // Show button after scrolling this much

// Helper function to get category style with theme awareness
const getCategoryStyle = (category) => {
    if (category === "cafe") return styles.cafeBadge;
    if (category === "restaurant") return styles.restaurantBadge;
    return styles.activityBadge;
};

const getCategoryText = (category) => {
    if (category === "cafe") return "Cafe";
    if (category === "restaurant") return "Restaurant";
    return "Activity";
};

// Memo-ize the POIListItem component to prevent unnecessary re-renders
const POIListItem = memo(({ item, userLocation, calculateDistance, themeStyles, isDarkMode, theme }) => {
    const [imageError, setImageError] = useState(false);
    const { handleButtonPress } = useButtonInteraction();

    // Use provided styles or fall back to defaults
    const styles = themeStyles || defaultStyles;
    
    // Use the pre-computed distance if available
    const poiDistance = item._distance !== undefined 
        ? item._distance 
        : calculateDistance(
            userLocation?.latitude || 0,
            userLocation?.longitude || 0,
            item.geometry?.location?.lat,
            item.geometry?.location?.lng
          );
    
    // Extract photo reference using proper path
    const photoReference = !imageError && item.photos?.[0]?.photo_reference;
    
    // Build photo URL following Google Places API requirements
    const imageUrl = photoReference
        ? `https://maps.googleapis.com/maps/api/place/photo?maxwidth=400&photoreference=${photoReference}&key=${GOOGLE_PLACES_API_KEY}`
        : null;
    
    const router = useRouter();
    const { updatePOILocationData } = useLocationContext();
    
    // Theme-aware icon color
    const placeholderIconColor = isDarkMode ? "#aaa" : "#888";
    const starIconColor = isDarkMode ? "#ffd700" : "#FFD700"; // Slightly adjusted for dark mode
    
    // Memoize the handler function to prevent recreation on each render
    const handleGetDirections = useCallback(() => {
        handleButtonPress(null, `Getting directions to ${item.name}`);
        console.log(`Get directions to: ${item.name}`);
        console.log(`Address: ${item.vicinity || "Address not available"}`);
        console.log(
          `Coordinates: ${item.geometry?.location?.lat}, ${item.geometry?.location?.lng}`
        );
        
        const name = item.name;
        const lat = item.geometry?.location?.lat;
        const lng = item.geometry?.location?.lng;
        
        // Update the context data without redirecting
        updatePOILocationData(name, lat, lng);
        
        // Navigate to a new page with the location details
        router.push({
            pathname: "/(tabs)/map",
            params: {
                name: item.name,
                lat: lat,
                lng: lng,
            },
        });
    }, [item.name, item.geometry?.location?.lat, item.geometry?.location?.lng, updatePOILocationData, router]);

    return (
        <View style={styles.poiItem}>
            {imageUrl && !imageError ? (
                <Image
                    source={{ uri: imageUrl }}
                    style={styles.poiImage}
                    resizeMode="cover"
                    testID="poi-image"
                    onError={(e) => {
                        console.log(`Image error for ${item.name}:`, e.nativeEvent.error);
                        setImageError(true);
                    }}
                    // Add fast image loading
                    fadeDuration={200}
                    progressiveRenderingEnabled={true}
                />
            ) : (
                <View style={styles.noImagePlaceholder}>
                    <Ionicons
                        name="image-outline"
                        size={40}
                        color={placeholderIconColor}
                    />
                    <Text style={styles.noImageText}>No Image Available</Text>
                </View>
            )}
            <View style={styles.poiContent}>
                <View style={styles.poiHeader}>
                    <Text style={styles.poiName} numberOfLines={1}>
                        {item.name}
                    </Text>
                    <Text style={styles.poiDistance}>
                        {poiDistance ? `${poiDistance.toFixed(1)} km` : ""}
                    </Text>
                </View>
                <Text style={styles.poiAddress} numberOfLines={2}>
                    {item.vicinity || "Address not available"}
                </Text>
                <View style={styles.categoryContainer}>
                    <View style={[styles.categoryBadge, getCategoryStyle(item.category)]}>
                        <Text style={styles.categoryText}>
                            {getCategoryText(item.category)}
                        </Text>
                    </View>
                    {item.rating && (
                        <View style={styles.ratingBadge}>
                            <Ionicons name="star" size={12} color={starIconColor} />
                            <Text style={styles.ratingText}>{item.rating}</Text>
                        </View>
                    )}
                </View>
                {/* Get Directions Button */}
                <TouchableOpacity
                    style={styles.directionsButton}
                    onPress={handleGetDirections}
                >
                    <Ionicons name="navigate" size={16} color="white" />
                    <Text style={styles.directionsButtonText}>Get Directions</Text>
                </TouchableOpacity>
            </View>
        </View>
    );
}, (prevProps, nextProps) => {
    // Custom comparison function for memo
    // Only re-render if the item ID changes or user location changes significantly
    const prevItem = prevProps.item;
    const nextItem = nextProps.item;
    
    // If place_id is different, always re-render
    if (prevItem.place_id !== nextItem.place_id) return false;
    
    // If user location changed significantly, re-render
    const prevLoc = prevProps.userLocation;
    const nextLoc = nextProps.userLocation;
    if (!prevLoc && nextLoc) return false;
    if (prevLoc && !nextLoc) return false;
    if (prevLoc && nextLoc) {
        const latDiff = Math.abs(prevLoc.latitude - nextLoc.latitude);
        const lngDiff = Math.abs(prevLoc.longitude - nextLoc.longitude);
        
        // Location change threshold that would make distance display noticeably different
        if (latDiff > 0.001 || lngDiff > 0.001) return false;
    }
    
    // If theme changes, re-render
    if (prevProps.isDarkMode !== nextProps.isDarkMode) return false;
    
    // If item's relevant properties changed, re-render
    return !(prevItem.name === nextItem.name &&
        prevItem.vicinity === nextItem.vicinity &&
        prevItem.rating === nextItem.rating &&
        prevItem._distance === nextItem._distance);
});

POIListItem.propTypes = {
    item: PropTypes.shape({
        name: PropTypes.string,
        place_id: PropTypes.string,
        vicinity: PropTypes.string,
        category: PropTypes.string,
        rating: PropTypes.number,
        _distance: PropTypes.number, // Pre-computed distance
        geometry: PropTypes.shape({
            location: PropTypes.shape({
                lat: PropTypes.number,
                lng: PropTypes.number
            })
        }),
        photos: PropTypes.arrayOf(
            PropTypes.shape({
                photo_reference: PropTypes.string
            })
        )
    }).isRequired,
    userLocation: PropTypes.shape({
        latitude: PropTypes.number,
        longitude: PropTypes.number
    }),
    calculateDistance: PropTypes.func.isRequired,
    themeStyles: PropTypes.object,
    isDarkMode: PropTypes.bool,
    theme: PropTypes.object,
};

// Function to extract a stable key for FlatList
const keyExtractor = item => item.uniqueKey || item.place_id || Math.random().toString();

const POIList = ({
  data,
  userLocation,
  isLoading,
  error,
  refreshing,
  onRefresh,
  calculateDistance,
  themeStyles,
  isDarkMode: propIsDarkMode,
  theme: propTheme,
}) => {
    // Get theme from context if not provided as props
    const themeContext = useContext(ThemeContext);
    const isDarkMode =
      propIsDarkMode !== undefined ? propIsDarkMode : themeContext?.isDarkMode;
    const theme = propTheme || themeContext?.theme;
    
    // Use provided styles or create theme-aware styles
    const styles =
      themeStyles ||
      (isDarkMode !== undefined
        ? createPOIListStyles({ isDarkMode, theme })
        : defaultStyles);
  
    // Theme-aware colors
    const loadingColor = isDarkMode
      ? theme?.buttonBackground || "#aa3355"
      : "#922338";
      
    // Add refs and state for scroll-to-top functionality
    const flatListRef = useRef(null);
    const [showScrollTopButton, setShowScrollTopButton] = useState(false);
    const scrollButtonOpacity = useRef(new Animated.Value(0)).current;
    
    // Memoize data to prevent unnecessary re-renders
    const memoizedData = useMemo(() => data, [data]);
    
    const { handleButtonPress } = useButtonInteraction();
    
    // Memoize renderItem function to prevent recreation on each render
    const renderItem = useCallback(({ item }) => (
        <POIListItem
            item={item}
            userLocation={userLocation}
            calculateDistance={calculateDistance}
            themeStyles={styles}
            isDarkMode={isDarkMode}
            theme={theme}
        />
    ), [userLocation, calculateDistance, styles, isDarkMode, theme]);
    
    // Ensure consistent item height for getItemLayout
    const getItemLayout = useCallback((_, index) => ({
        length: ITEM_HEIGHT,
        offset: ITEM_HEIGHT * index,
        index,
    }), []);
    
    // Handle scroll event to show/hide scroll-to-top button
    const handleScroll = useCallback((event) => {
        const scrollY = event.nativeEvent.contentOffset.y;
        
        // Show button when scrolled down enough
        if (scrollY > SCROLL_THRESHOLD && !showScrollTopButton) {
            setShowScrollTopButton(true);
            Animated.timing(scrollButtonOpacity, {
                toValue: 1,
                duration: 200,
                useNativeDriver: true
            }).start();
        } 
        // Hide button when close to the top
        else if (scrollY <= SCROLL_THRESHOLD && showScrollTopButton) {
            Animated.timing(scrollButtonOpacity, {
                toValue: 0,
                duration: 200,
                useNativeDriver: true
            }).start(() => {
                setShowScrollTopButton(false);
            });
        }
    }, [showScrollTopButton, scrollButtonOpacity]);
    
    // Scroll to top function
    const scrollToTop = useCallback(() => {
        if (flatListRef.current) {
            flatListRef.current.scrollToOffset({ offset: 0, animated: true });
        }
    }, []);

    if (isLoading && !refreshing) {
        return (
            <View style={styles.loadingContainer}>
                <ActivityIndicator
                  size="large"
                  color={loadingColor}
                  testID="activity-indicator"
                />
                <Text style={styles.loadingText}>Loading places...</Text>
            </View>
        );
    }
    
    if (error && (!data || data.length === 0)) {
        return (
            <View style={styles.noResultsContainer}>
                <Text style={styles.errorText}>{error ?? "An error occurred"}</Text>
                <TouchableOpacity
                    style={[styles.retryButton, { marginTop: 20 }]}
                    onPress={() => {
                        handleButtonPress(null, 'Retrying to load places');
                        onRefresh();
                    }}
                >
                    <Ionicons name="refresh" size={18} color="white" />
                    <Text style={styles.buttonText}>Retry</Text>
                </TouchableOpacity>
            </View>
        );
    }
    
    if (!data || data.length === 0) {
        return (
            <View style={styles.noResultsContainer}>
                <Text style={styles.noResultsText}>
                    No places found with the current filters.
                    Try adjusting your filter settings or increasing the distance.
                </Text>
            </View>
        );
    }
    
    console.log(`POIList rendering ${data.length} items`);
    
    return (
        <View style={{ flex: 1 }}>
            <FlatList
                testID="poi-flatlist"
                ref={flatListRef}
                data={memoizedData}
                renderItem={renderItem}
                keyExtractor={keyExtractor}
                contentContainerStyle={styles.listContainer}
                showsVerticalScrollIndicator={true}
                
                // Performance optimization settings
                initialNumToRender={5}
                maxToRenderPerBatch={3}
                updateCellsBatchingPeriod={100}
                windowSize={5}
                removeClippedSubviews={Platform.OS !== 'web'} // This helps on mobile
                
                // Use consistent item heights for smoother scrolling
                getItemLayout={getItemLayout}
                
                // Add scroll event listener
                onScroll={handleScroll}
                scrollEventThrottle={16}
                
                // Additional optimizations
                maintainVisibleContentPosition={{
                    minIndexForVisible: 0,
                    autoscrollToTopThreshold: 3
                }}
                
                refreshControl={
                    <RefreshControl
                        refreshing={refreshing}
                        onRefresh={onRefresh}
                        colors={[loadingColor]}
                        tintColor={loadingColor}
                        progressBackgroundColor={isDarkMode ? "#333" : "#fff"}
                    />
                }
                
                // Progressive loading indicator
                ListFooterComponent={
                    data.length > 20 ? (
                        <View style={styles.footerContainer}>
                            <Text style={styles.footerText}>
                                {`Showing ${data.length} places`}
                            </Text>
                        </View>
                    ) : null
                }
            />
            
            {/* Scroll to top button */}
            {showScrollTopButton && (
                <Animated.View style={[
                    styles.scrollTopButton, 
                    { opacity: scrollButtonOpacity }
                ]}>
                    <TouchableOpacity
                        onPress={scrollToTop}
                        style={styles.scrollTopButtonTouchable}
                        activeOpacity={0.8}
                    >
                        <Ionicons name="arrow-up" size={24} color="white" />
                    </TouchableOpacity>
                </Animated.View>
            )}
        </View>
    );
};

POIList.propTypes = {
  data: PropTypes.array,
  userLocation: PropTypes.shape({
    latitude: PropTypes.number,
    longitude: PropTypes.number,
  }),
  isLoading: PropTypes.bool,
  error: PropTypes.string,
  refreshing: PropTypes.bool,
  onRefresh: PropTypes.func.isRequired,
  calculateDistance: PropTypes.func.isRequired,
  themeStyles: PropTypes.object,
  isDarkMode: PropTypes.bool,
  theme: PropTypes.object,
};

export default memo(POIList);
