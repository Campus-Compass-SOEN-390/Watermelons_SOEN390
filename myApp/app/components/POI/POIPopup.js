import React, { useContext } from "react";
import { View, Text, TouchableOpacity, StyleSheet } from "react-native";
import { MaterialIcons } from "@expo/vector-icons";
import PropTypes from "prop-types";
import { ThemeContext } from "../../context/ThemeContext"; // Update this path if needed

const POIPopup = ({
  poi = null,
  distance = null,
  onClose,
  onGetDirections,
  isDarkMode: propIsDarkMode,
  theme: propTheme,
}) => {
  // Get theme from context
  const themeContext = useContext(ThemeContext);

  if (!poi) return null;

  const isDarkMode =
    propIsDarkMode !== undefined ? propIsDarkMode : themeContext?.isDarkMode;
  const theme = propTheme || themeContext?.theme;

  // Create theme-aware styles
  const themedStyles = createThemedStyles(isDarkMode, theme);

  // Format distance in a readable way
  const formatDistance = (meters) => {
    if (meters < 1000) {
      return `${Math.round(meters)} m`;
    } else {
      return `${(meters / 1000).toFixed(1)} km`;
    }
  };

  return (
    <View
      style={themedStyles.container}
      testID="poi-popup"
      accessible={true}
      accessibilityLabel="Point of interest information"
    >
      {/* Close button */}
      <TouchableOpacity
        style={themedStyles.closeButton}
        onPress={onClose}
        testID="close-button"
        accessible={true}
        accessibilityLabel="Close point of interest information"
        accessibilityRole="button"
      >
        <MaterialIcons
          name="close"
          size={20}
          color={isDarkMode ? theme?.text || "#bbb" : "#666"}
        />
      </TouchableOpacity>

      {/* POI info */}
      <Text style={themedStyles.name}>{poi.name}</Text>
      <Text style={themedStyles.detail}>
        {poi.rating
          ? `â˜… ${poi.rating} (${poi.user_ratings_total || 0} ratings)`
          : "No ratings"}
      </Text>
      <Text style={themedStyles.detail}>{poi.vicinity || ""}</Text>
      <Text style={themedStyles.detail}>
        {distance ? `Distance: ${formatDistance(distance)}` : ""}
      </Text>

      {/* Get directions button */}
      <TouchableOpacity
        style={themedStyles.directionsButton}
        onPress={onGetDirections}
        testID="directions-button"
        accessible={true}
        accessibilityLabel="Get directions to this location"
        accessibilityRole="button"
      >
        <MaterialIcons name="directions" size={16} color="white" />
        <Text style={themedStyles.directionsText}>Get Directions</Text>
      </TouchableOpacity>
    </View>
  );
};

// Define prop types
POIPopup.propTypes = {
  poi: PropTypes.shape({
    name: PropTypes.string.isRequired,
    rating: PropTypes.number,
    user_ratings_total: PropTypes.number,
    vicinity: PropTypes.string,
  }),
  distance: PropTypes.number,
  onClose: PropTypes.func.isRequired,
  onGetDirections: PropTypes.func.isRequired,
  isDarkMode: PropTypes.bool,
  theme: PropTypes.object,
};

// Create theme-aware styles
const createThemedStyles = (isDarkMode, theme) => {
  const backgroundColor = isDarkMode
    ? theme?.cardBackground || "#2a2a2a"
    : "white";
  const textColor = isDarkMode ? theme?.text || "#fff" : "#000";
  const detailColor = isDarkMode ? theme?.subText || "#aaa" : "#666";
  const buttonColor = theme?.buttonBackground || "#922338";

  return StyleSheet.create({
    container: {
      backgroundColor: backgroundColor,
      borderRadius: 8,
      padding: 12,
      width: 220,
      shadowColor: "#000",
      shadowOffset: { width: 0, height: 2 },
      shadowOpacity: isDarkMode ? 0.5 : 0.25,
      shadowRadius: 3.84,
      elevation: 5,
      borderWidth: isDarkMode ? 1 : 0,
      borderColor: isDarkMode ? "#444" : "transparent",
    },
    closeButton: {
      position: "absolute",
      right: 8,
      top: 8,
      zIndex: 1,
    },
    name: {
      fontWeight: "bold",
      fontSize: 16,
      marginBottom: 4,
      paddingRight: 24,
      color: textColor,
    },
    detail: {
      fontSize: 12,
      color: detailColor,
      marginBottom: 2,
    },
    directionsButton: {
      backgroundColor: buttonColor,
      borderRadius: 4,
      flexDirection: "row",
      alignItems: "center",
      justifyContent: "center",
      paddingVertical: 8,
      marginTop: 8,
    },
    directionsText: {
      color: "white",
      fontWeight: "bold",
      marginLeft: 4,
    },
  });
};

export default POIPopup;
