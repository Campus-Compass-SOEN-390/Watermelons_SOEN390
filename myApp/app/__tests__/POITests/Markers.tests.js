import React from 'react';
import { render } from '@testing-library/react-native';
import { CoffeeMarker, RestaurantMarker, ActivityMarker } from '../../components/POI/Markers';
import { MaterialCommunityIcons } from '@expo/vector-icons';

// Mock the @expo/vector-icons
jest.mock('@expo/vector-icons', () => {
  const { View } = require('react-native');
  return {
    MaterialCommunityIcons: jest.fn(({ name, size, color }) => (
      <View testID={`icon-${name}`} size={size} color={color} />
    )),
  };
});

describe('Marker Components', () => {
  beforeEach(() => {
    // Clear mock calls between tests
    MaterialCommunityIcons.mockClear();
  });

  describe('CoffeeMarker', () => {
    it('renders correctly', () => {
      const { getByTestId } = render(<CoffeeMarker />);
      expect(getByTestId('icon-coffee')).toBeTruthy();
    });

    it('uses the correct icon properties', () => {
      render(<CoffeeMarker />);
      expect(MaterialCommunityIcons).toHaveBeenCalledWith(
        expect.objectContaining({
          name: 'coffee',
          size: 24,
          color: 'black'
        }),
        expect.anything()
      );
    });

    it('has the correct container styling', () => {
      const { UNSAFE_getByType } = render(<CoffeeMarker />);
      const view = UNSAFE_getByType('View');
      expect(view.props.style).toEqual({
        backgroundColor: 'white',
        padding: 5,
        borderRadius: 5
      });
    });
  });

  describe('RestaurantMarker', () => {
    it('renders correctly', () => {
      const { getByTestId } = render(<RestaurantMarker />);
      expect(getByTestId('icon-silverware-fork-knife')).toBeTruthy();
    });

    it('uses the correct icon properties', () => {
      render(<RestaurantMarker />);
      expect(MaterialCommunityIcons).toHaveBeenCalledWith(
        expect.objectContaining({
          name: 'silverware-fork-knife',
          size: 24,
          color: 'orange'
        }),
        expect.anything()
      );
    });

    it('has the correct container styling', () => {
      const { UNSAFE_getByType } = render(<RestaurantMarker />);
      const view = UNSAFE_getByType('View');
      expect(view.props.style).toEqual({
        backgroundColor: 'white',
        padding: 5,
        borderRadius: 5
      });
    });
  });

  describe('ActivityMarker', () => {
    it('renders correctly', () => {
      const { getByTestId } = render(<ActivityMarker />);
      expect(getByTestId('icon-run')).toBeTruthy();
    });

    it('uses the correct icon properties', () => {
      render(<ActivityMarker />);
      expect(MaterialCommunityIcons).toHaveBeenCalledWith(
        expect.objectContaining({
          name: 'run',
          size: 20,
          color: 'green'
        }),
        expect.anything()
      );
    });

    it('has the correct container styling', () => {
      const { UNSAFE_getByType } = render(<ActivityMarker />);
      const view = UNSAFE_getByType('View');
      expect(view.props.style).toEqual({
        backgroundColor: 'white',
        padding: 5,
        borderRadius: 5
      });
    });
  });

  describe('Snapshots', () => {
    it('CoffeeMarker matches snapshot', () => {
      const tree = render(<CoffeeMarker />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('RestaurantMarker matches snapshot', () => {
      const tree = render(<RestaurantMarker />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('ActivityMarker matches snapshot', () => {
      const tree = render(<ActivityMarker />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
