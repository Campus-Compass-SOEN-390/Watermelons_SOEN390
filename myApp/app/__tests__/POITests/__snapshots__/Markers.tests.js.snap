// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marker Components Snapshots ActivityMarker matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "borderRadius": 5,
      "padding": 5,
    }
  }
>
  <View
    color="green"
    size={20}
    testID="icon-run"
  />
</View>
`;

exports[`Marker Components Snapshots CoffeeMarker matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "borderRadius": 5,
      "padding": 5,
    }
  }
>
  <View
    color="black"
    size={24}
    testID="icon-coffee"
  />
</View>
`;

exports[`Marker Components Snapshots RestaurantMarker matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "borderRadius": 5,
      "padding": 5,
    }
  }
>
  <View
    color="orange"
    size={24}
    testID="icon-silverware-fork-knife"
  />
</View>
`;
